### ISSUE TITLE
Running out of sockets using router-dealer pattern

### ISSUE THREAD

**Issue by MVJMBoeren** [zeromq/libzmq#2823]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description
Running out of sockets while trying to connect an unreachable ip address, using router-dealer zmq pattern.
After 2022 attempts the next attempt to create a zmq_socket returns nullptr.
(See also #2822).
The first 1000 sockets are destroyed as expected. The sockets above 1000 aren't destroyed anymore leading to a problem with opening new sockets.

# Environment
Windows 

* libzmq version (commit hash if unreleased): 
4.2.0
* OS: Windows 7

# Minimal test code / Steps to reproduce the issue

```
TEST_F(ZmqCTest, RouterDealerTest)
{
    auto context = zmq_ctx_new();
    std::string request = "Hello!";
    int i(0);

    auto routerSocket = zmq_socket(context, ZMQ_ROUTER);
    auto dealerSocket = zmq_socket(context, ZMQ_DEALER);
    
    static const char* inProcName = "inproc://#1";
    zmq_bind(routerSocket, inProcName);

    zmq_connect(dealerSocket, "tcp://192.168.10.234:5559"); //Unreachable IP

    auto runCallback = [&]()
    {
        zmq_proxy(routerSocket, dealerSocket, nullptr);
        zmq_close(routerSocket);
        zmq_close(dealerSocket);
    };

    auto thread = std::make_unique<std::thread>(runCallback);

    while (true)
    {
        i++;
        std::cout << "Attempt " << i << std::endl;

        auto socket = zmq_socket(context, ZMQ_REQ);
        if (socket == nullptr)
        {
            throw std::runtime_error("Running out of sockets?");
        }

        zmq_connect(socket, inProcName);

        zmq_send(socket, request.data(), request.size(), 0);

        zmq_close(socket);
    }
    zmq_ctx_destroy(context);
    thread->join();
}
```

# What's the actual result? (include assertion message & call stack if applicable)
We're not able to open new sockets.

# What's the expected result?
We should be able to open new sockets as long as we correctly closed the already used ones.

**Comment by bluca** [comment]
TL;DR - known "problem" of deciding whether to block or drop messages or have potential unlimited memory usage when a peer is unresponsive

The problem is that the sockets are never really and fully deallocated, and the program hits the per-process file descriptor limit, that's why it cannot allocate new ones. If you try to change it (uname -n) you'll see that the counter changes accordingly.

There are 2 problems.

1) Given the dealer never connects, the proxy will only forward messages until it hits the HWM. Then ZMQ_POLLOUT will be false since the pipe is full, and the proxy will not foward anymore. As a side effect, the router command state machine will not run anymore since there's no code to exercise it (eg: send/recv/get/setsockopt) that runs. If you change the SNDHWM of the dealer, again you'll see the counter change.
2) But even if that is fixed, for example by adding a dummy getsockopt in the proxy implementation so that the commands get processed, for inproc there is no I/O thread involvement so all the messages are in-flight in the pipe, and never received, given they cannot be forwarded. This means that even if the router receives the pipe_term command, its pipe's state machine then gets stuck in the waiting_for_delimiter state, since the pipe is full of unread messages. Given they will never be read, it's stuck and never sends the pipe_term_ack back to the reaper thread, which cannot deallocate the socket and free the file descriptor.

If the HWM situation gets resolved, then the library recovers and starts to work again. To prove this, try to have another program bind to the port that the dealer tries to connect to, and you'll see the program recovers. You'll need to change to block rather than assert when zmq_socket returns an error, eg:

```
   do {
        socket = zmq_socket(context, ZMQ_REQ);
   } while (socket == nullptr && errno == EMFILE);
```

Workarounds:

- set the SND_HWM to 0 (unlimited) in the dealer. This means potential unbounded memory usage, and depending on the application and use case that might not work.
- use socket events in the proxy thread in the application, and don't actually call zmq_proxy until you get a _CONNECTED event, and in the meanwhile drain the router and discard the messages. That assumes the application protocol is "lossy".
- switch to TCP rather than inproc - TCP uses an I/O thread that can handle the messages, so pipe terminations won't be held back. This again means dropping messages though once the dealer HWM is reached - it's just hidden due to the way that router by default drops once in mute state.

So it's a matter of deciding what's best for your application - block, drop, or potential unlimited memory usage.

I'm afraid there's no "magic" solution here - it's a well known networking problem.

**Comment by bluca** [comment]
Also - create and bind the router and dealer in the callback that runs in the other thread, as sockets are not thread safe

**Comment by oponomarov** [comment]
@bluca:
Thanks for the proposed workarounds.  As any workaround they help mitigate the problem and not to solve it. Actual solving still must be implemented in the library itself. And consideration about this is very simple: pipe is blocked because it is full of messages which are not relevant to any client since all sockets through which corresponding messages could be accessed were already closed. So implemented properly all messages related to sockets which were closed by client must be automatically removed from zmq pipe. This will solve the problem and this will be "magic" solution I am looking for although it can require significant resources to implement this.

Let's consider your workarounds.
Option 1 apparently is not really applicable since if the connection is absent in long run we will have problems with memory allocations so it only defers the moment when client code will face the problem.
Option 3 I suppose has performance drawbacks compared to using 'inproc' sockets, right?
Option 2 looks to be the most applicable one.  What I don't understand is how I can "in the meanwhile drain the router and discard the messages". Can you explain this in more details? As you can see in the code snippet in separate thread function "zmq_proxy" is called and it executes till some error or destroying the ZMQ context. I do't understand what can be done to "drop the messages". Please explain.

Thank you in advance!

**Comment by oponomarov** [comment]
@bluca:
I have also one concern about option 2 with monitoring the status of ZMQ socket.  It's about problem described in https://github.com/zeromq/libzmq/issues/1552. In that ticket even after receiving CONNECTED event for the socket immediate request to send something via that socket fails which is really strange and opens the question about how these events can be in theory used, how client can rely on them. If after getting CONNECTED event you still have to wait some unknown time before being able to send message this spoils the whole idea with events.
Thank you in advance.
